### PR TITLE
feat(enterprise): add hostname field when reporting a configuration version

### PR DESCRIPTION
**Pull request by jdrouet**
To avoid reporting several time the same configuration version for a given host, we need to report the current hostname. This PR just adds the `hostname` field to the reported data.

Related to https://datadoghq.atlassian.net/browse/OPB-80

Signed-off-by: Jeremie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

**Comment by spencergilbert** [comment]
Without totally being aware of the use-case here, this would likely return the name of Vector's pod when running in kubernetes.

**Comment by jdrouet** [comment]
> Without totally being aware of the use-case here, this would likely return the name of Vector's pod when running in kubernetes.

@spencergilbert is there a way to get that from the environment variable? I've the impression that it's the same as the `HOSTNAME` 🤔 

The goal of this `hostname` field is just to identify an instance of vector with it's version when reporting a new configuration to the obs pipeline

**Comment by spencergilbert** [comment]
If they're using the default Helm chart - we inject the actual node's name as a variable: https://github.com/vectordotdev/helm-charts/blob/develop/charts/vector/templates/_pod.tpl#L53-L56

This seems to be the generally suggested method, otherwise there's API calls to kubernetes infrastructure or weird attempts at parsing machine-id.

File: src/config/enterprise.rs [hunk 1]
```diff
@@ -601,6 +604,13 @@ pub(crate) fn report_configuration(
             opts,
         } = metadata;
 
+        // Get the machine hostname to avoid reporting the same configuration several time
+        // for a given instance of vector.
+        // If running on kubernetes, we take the pod name instead.
+        let hostname = env::var("VECTOR_SELF_POD_NAME").unwrap_or_else(|_| {
```
Comment (spencergilbert) [inline src/config/enterprise.rs:610]:
```suggestion
        let hostname = env::var("VECTOR_SELF_NODE_NAME").unwrap_or_else(|_| {
```

File: src/config/enterprise.rs [hunk 2]
```diff
@@ -601,6 +604,13 @@ pub(crate) fn report_configuration(
...
+            crate::get_hostname().expect("Couldn't read the current hostname.")
```
Comment (spencergilbert) [inline src/config/enterprise.rs:611]: No strong opinion on the fallback here as it seems like a product/business decision, which I'm fairly distant from these days.

**Review by spencergilbert — CHANGES_REQUESTED** [review]

File: src/config/enterprise.rs (hunk 1 above)
Comment (jdrouet) [reply src/config/enterprise.rs:610]: 🤔 didn't you say pod name?

Comment (spencergilbert) [reply src/config/enterprise.rs:610]: Looks like I said node and linked to node. Node name will be the name of the physical/virtual machine any pod is running on pod name would be the equivalent of `get_hostname`

File: src/config/enterprise.rs [hunk 3]
```diff
@@ -622,6 +625,13 @@ pub(crate) fn report_configuration(
             opts,
         } = metadata;
 
+        // Get the machine hostname to avoid reporting the same configuration several time
+        // for a given instance of vector.
+        // If running on kubernetes, we take the pod name instead.
+        let hostname = env::var("VECTOR_SELF_POD_NAME")
+            .ok()
+            .or_else(|| crate::get_hostname().ok());
```
[the remark this answers is not in the document]
Comment (jdrouet) [reply src/config/enterprise.rs]: Using the pod name was a proposal from @spencergilbert, I don't really have an opinion about it.
I removed it.

**Review by jszwedko — APPROVED** [review]

**Comment by jszwedko** [comment]
Just noting that this could still result in publishing without a hostname if the hostname couldn't be detected. Is that a hard requirement for OP? If so, we could probably error if the hostname couldn't be determined. We have #14250 to cover this more generally within Vector.

**Comment by jdrouet** [comment]
This feature has been put on hold for now so we can close this for now I guess.

**Comment by jszwedko** [comment]
Closing per https://github.com/vectordotdev/vector/pull/13703#issuecomment-1271310613

**Pull request closed without merging**